### PR TITLE
test(flat-table): refine test suites

### PR DESCRIPTION
**Pull request by sianford**
### Proposed behaviour

- Removes duplicate and unnecessary tests from the `FlatTable` Playwright file.
- Adds/removes tests in the `FlatTable` Jest file. 
- Consolidates/disables Chromatic snapshots to be more efficient.

### Current behaviour

- There are a number of duplicated tests in the Playwright file which are covered in other tools.
- There are a number of tests missing or required that are better placed in the Jest file.
- There are extra snapshots being taken which aren't currently required.

### Checklist

- [ ] Commits follow our style guide
- [ ] Related issues linked in commit messages if required
- [ ] Screenshots are included in the PR if useful
- [ ] All themes are supported if required
- [ ] Unit tests added or updated if required
- [ ] Playwright automation tests added or updated if required
- [ ] Storybook added or updated if required
- [ ] Translations added or updated (including creating or amending translation keys table in storybook) if required
- [ ] Typescript `d.ts` file added or updated if required
- [ ] Related docs have been updated if required

#### QA

- [ ] Tested in Storybook
- [ ] Add new Playwright test coverage if required
- [ ] Carbon implementation matches Design System/designs
- [ ] UI Tests GitHub check reviewed if required

### Testing instructions

Ensure Playwright tests pass and Chromatic snapshots are as expected.

### CODE REVIEW

File: src/components/flat-table/flat-table-expandable.stories.tsx [hunk 1]
```diff
@@ -60,6 +60,9 @@ type SubRowsShapeChildrenOnlySelectableStoryKey =
 const meta: Meta<typeof FlatTable> = {
   title: "Flat Table/Expandable",
   component: FlatTable,
+  parameters: {
```
Comment (paulrobinson88) [inline src/components/flat-table/flat-table-expandable.stories.tsx:63]: **comment(non-blocking)**: It looks like we're currently capturing the `sage` and `none` theme, we could alter this section to define only the sage theme 
```
  parameters: {
    themeProvider: { chromatic: { theme: "sage" } },
    chromatic: { disableSnapshot: false },
```

File: src/components/flat-table/flat-table-test.stories.tsx [hunk 2]
```diff
@@ -45,7 +46,7 @@ export default {
   parameters: {
     info: { disable: true },
     chromatic: {
-      disableSnapshot: true,
+      disableSnapshot: false,
```
Comment (paulrobinson88) [inline src/components/flat-table/flat-table-test.stories.tsx:50]: Similar to the `flat-table-expandable.stories.tsx` we could define the theme here at the top rather than specifying per story like it is in this file.

File: src/components/flat-table/flat-table.interaction.stories.tsx
Comment (paulrobinson88) [inline src/components/flat-table/flat-table.interaction.stories.tsx:1]: **question: (non-blocking)** Is the purpose of the interactions mainly to capture focus styling in these different circumstances? Just wonder if we could have a smaller file based on just doing this whereas there are a large number of interactions which wouldn't totally be necessary unless we needed to capture functionality that isn't covered in Jest or Playwright.

File: src/components/flat-table/flat-table.pw.tsx [hunk 3]
```diff
@@ -22,17 +19,13 @@ import {
...
-  FlatTableParentSubrowSelectableComponent,
   FlatTableChildSubrowSelectableComponent,
   FlatTablePartiallySelectedOrHighlightedRows,
   FlatTableDraggableComponent,
   FlatTablePagerStickyHeaderComponent,
-  FlatTableCheckboxComponent,
-  FlatTableFirstColumnHasRowspan,
   FlatTableLastColumnHasRowspan,
   FlatTableWithStickyColumn,
 } from "./components.test-pw";
```
Comment (paulrobinson88) [inline src/components/flat-table/flat-table.pw.tsx:27]: **Question:** As we have removed a number of these patterns, are these removed lines used elsewhere from the `components.test-pw` file? If not we could maybe remove these from that file.

File: src/components/flat-table/flat-table.interaction.stories.tsx
Comment (sianford) [reply src/components/flat-table/flat-table.interaction.stories.tsx:1]: The focus is really weird on certain components within FlatTable. I tried having a story in Test for just the focus states but I couldn't apply a pseudo focus to it. In the Chrome DevTools, the only way I could force a focus was with `focus-hidden` ticked on those elements. Standard `focus` didn't focus it. So the only way to focus a component was by using Interactions to tab to it/click on it. As such, I removed a load of the focus tests out of Playwright.

File: src/components/flat-table/flat-table-test.stories.tsx [hunk 4]
```diff
@@ -453,56 +386,55 @@ export const ExpandableWithActionPopover = () => {
           <FlatTableRowHeader>Name</FlatTableRowHeader>
           <FlatTableHeader>Location</FlatTableHeader>
           <FlatTableHeader>Relationship Status</FlatTableHeader>
-          <FlatTableHeader>Actions</FlatTableHeader>
+          <FlatTableHeader>Dependents</FlatTableHeader>
```
[the remark this answers is not in the document]
Comment (sianford) [reply src/components/flat-table/flat-table-test.stories.tsx]: Not relevant for this test. This test just shows different nodes in an expandable row.

File: src/components/flat-table/flat-table.interaction.stories.tsx
Comment (sianford) [reply src/components/flat-table/flat-table.interaction.stories.tsx:1]: For functionality like sorting columns in ascending/descending order, I combined several tests aspects in this file. For instance, I left keyboard functionality tests in Playwright but used mouseclick in the Interactions to test the mouseclick but also combine the Chromatic snapshots from the separate main stories into this one story to test the change of direction of the icon on the sorted column. Similarly, the for selectable and expandable rows, I removed several Playwright tests, mainly around mouse clicks and checking of styles, and snapshots from the separate main stories and combined several aspects of testing in the single story in this file to reduce snapshots and also avoid doing the convoluted ones in Playwright.

File: src/components/flat-table/flat-table.pw.tsx (hunk 3 above)
Comment (sianford) [reply src/components/flat-table/flat-table.pw.tsx:27]: Good spot! I had forgotten to remove them from that file.

File: src/components/flat-table/flat-table.pw.tsx [hunk 5]
```diff
@@ -2300,124 +1514,135 @@ test.describe("Prop tests", () => {
...
-      for await (const i of indexes(4)) {
-        await expect(
-          flatTableCheckboxAsProp(page, i, asPropVal).first(),
-        ).toHaveAttribute("data-element", "flat-table-checkbox-header");
-      }
-    });
-  });
-
-  test(`should apply correct focus styling when wrapper is focused via keyboard`, async ({
+test.describe("Rounded corner tests", () => {
```
Comment (edleeks87) [inline src/components/flat-table/flat-table.pw.tsx]: Comment: this seems like something we could capture in chromaitc

File: src/components/flat-table/flat-table.pw.tsx
Comment (edleeks87) [inline src/components/flat-table/flat-table.pw.tsx:1]: Comment: I don't think we need any of the `Prop tests` here, they're all covered in unit and most will also be covered by chromatic albeit not directly in some cases

File: src/components/flat-table/flat-table.pw.tsx [hunk 6]
```diff
@@ -2442,7 +1667,7 @@ test.describe("Accessibility tests", () => {
   });
 
   [...sizes].forEach(([sizeName]) => {
```
Comment (edleeks87) [inline src/components/flat-table/flat-table.pw.tsx]: Comment: I'm not sure how the size prop affects accessibility

File: src/components/flat-table/flat-table.pw.tsx [hunk 7]
```diff
@@ -2452,70 +1677,6 @@ test.describe("Accessibility tests", () => {
...
-  test(`should render with rowSpan set to make header cells span 2 rows for accessibility tests`, async ({
-    mount,
-    page,
-  }) => {
-    await mount(<FlatTableSpanComponent width="500px" />);
-
-    await checkAccessibility(page);
-  });
-
   test(`should render with multiple sticky row headers for accessibility tests`, async ({
```
Comment (edleeks87) [inline src/components/flat-table/flat-table.pw.tsx]: comment: not sure how sticky columns affect accessibility

File: src/components/flat-table/flat-table.pw.tsx [hunk 8]
```diff
@@ -2526,33 +1687,6 @@ test.describe("Accessibility tests", () => {
...
-  test(`should render with footer for accessibility tests`, async ({
-    mount,
-    page,
-  }) => {
-    await mount(<FlatTableFooterComponent />);
-
-    await checkAccessibility(page);
-  });
-
   test(`should render with custom padding for accessibility tests`, async ({
```
Comment (edleeks87) [inline src/components/flat-table/flat-table.pw.tsx:1395]: comment: not sure how padding affects accessibility

File: src/components/flat-table/flat-table.interaction.stories.tsx [hunk 9]
```diff
@@ -13,21 +35,430 @@ export default {
...
+    },
   },
 };
 
-export const FlatTableFocused: Story = {
-  ...DefaultStory,
+const FlatTableFocusStatesComponent = () => {
+  const [recordsRange, setRecordsRange] = useState({ start: 0, end: 10 });
+  const [currentPage, setCurrentPage] = useState(1);
+  const rows = [
```
Comment (edleeks87) [inline src/components/flat-table/flat-table.interaction.stories.tsx]: suggestion: I don't think we need this many rows and cells etc this story is extremely complex for what it's trying to do. You could also iterate over some data like so and then you only need to define the JSX once etc

```
const data = [
  { name: "John Doe", location: "London", status: "Single" },
  { name: "Jane Doe", location: "York", status: "Married" },
  { name: "John Smith", location: "Edinburgh", status: "Single" },
  { name: "Jane Smith", location: "Newcastle", status: "Married" },
];

const renderRows = () => {
  const { start, end } = recordsRange;

  const slice = start < 0
    ? data
    : data.slice(start, end > data.length ? data.length : end);

  return slice.map(({ name, location, status, count }, index) => (
    <FlatTableRow key={index}>
      <FlatTableCell>{name}</FlatTableCell>
      <FlatTableCell>{location}</FlatTableCell>
      <FlatTableCell>{status}</FlatTableCell>
=    </FlatTableRow>
  ));
};

File: src/components/flat-table/flat-table.interaction.stories.tsx [hunk 10]
```diff
@@ -13,21 +35,430 @@ export default {
...
+  };
+
+  const composedHighlightableStory = composeStory(WithHighlightableRows, meta);
+
+  return (
+    <main>
+      <Box mb={2}>
+        <h2> Default</h2>
+        <Box mb={4}>
+          <FlatTable data-role="default-table" title="Default Flat Table">
```
Comment (edleeks87) [inline src/components/flat-table/flat-table.interaction.stories.tsx]: Question: what does this first table do as it's not focusable?

File: src/components/flat-table/flat-table.interaction.stories.tsx [hunk 11]
```diff
@@ -13,21 +35,430 @@ export default {
...
     themeProvider: { chromatic: { theme: "sage" } },
+    chromatic: {
+      disableSnapshot: false,
+    },
   },
 };
 
-export const FlatTableFocused: Story = {
-  ...DefaultStory,
+const FlatTableFocusStatesComponent = () => {
```
Comment (edleeks87) [inline src/components/flat-table/flat-table.interaction.stories.tsx:44]: comment: I think it's a good idea to do some consolidation here but we can really reduce down the complexity of it, we only need a few rows and columns etc

File: src/components/flat-table/flat-table.interaction.stories.tsx [hunk 12]
```diff
@@ -13,21 +35,430 @@ export default {
...
+          <Box data-role="highlightable-rows-example">
+            {composedHighlightableStory({ colorTheme: "dark" })}
+          </Box>
+        </Box>
+      </Box>
+    </main>
+  );
+};
+
+export const FlatTableFocusStates: Story = {
```
Comment (edleeks87) [inline src/components/flat-table/flat-table.interaction.stories.tsx:211]: Comment: similar to above and to what Paul has said I'd prefer to use `pseudo` here as this is very complex and involves a lot of interactions etc

File: src/components/flat-table/flat-table.interaction.stories.tsx [hunk 13]
```diff
@@ -37,20 +468,229 @@ export const FlatTableFocused: Story = {
...
+  decorators: [
+    (StoryToRender) => (
+      <DefaultDecorator>
+        <StoryToRender />
+      </DefaultDecorator>
+    ),
+  ],
+};
+
+const composedSelectableStory = composeStory(WithSelectableRows, meta);
```
Comment (edleeks87) [inline src/components/flat-table/flat-table.interaction.stories.tsx:365]: comment: we don't need these imo, we've unit/pw tested so all we need is to capture the state via setting a prop in chromatic

File: src/components/flat-table/flat-table.pw.tsx (hunk 8 above)
Comment (sianford) [reply src/components/flat-table/flat-table.pw.tsx:1395]: I left this in in case it highlighted an issue with the amount of white space between elements.

File: src/components/flat-table/flat-table.interaction.stories.tsx (hunk 12 above)
Comment (sianford) [reply src/components/flat-table/flat-table.interaction.stories.tsx:211]: Pseudo focus was not working when I tried. Checked with Dan and he thought that was may be why he didn't use pseudo focus in the original test cases.

File: src/components/flat-table/flat-table.interaction.stories.tsx (hunk 13 above)
Comment (sianford) [reply src/components/flat-table/flat-table.interaction.stories.tsx:365]: As above, there was an issue with pseudo focus so had to use an interaction instead.

File: src/components/flat-table/flat-table.interaction.stories.tsx (hunk 10 above)
Comment (sianford) [reply src/components/flat-table/flat-table.interaction.stories.tsx]: The table is focusable. The test tabs to it (lines 405 and 406).

File: src/components/flat-table/flat-table.interaction.stories.tsx (hunk 9 above)
Comment (sianford) [reply src/components/flat-table/flat-table.interaction.stories.tsx]: Amended

File: src/components/flat-table/flat-table.interaction.stories.tsx
Comment (sianford) [reply src/components/flat-table/flat-table.interaction.stories.tsx:1]: I'll log another ticket for followup for this part may be for someone else ( ;) ) to find a way to focus the required elements.

File: src/components/flat-table/flat-table.interaction.stories.tsx (hunk 11 above)
Comment (sianford) [reply src/components/flat-table/flat-table.interaction.stories.tsx:44]: I've removed some examples and simplified the structure, as recommended.

File: src/components/flat-table/flat-table.interaction.stories.tsx (hunk 9 above)
Comment (sianford) [reply src/components/flat-table/flat-table.interaction.stories.tsx]: Amended.

File: src/components/flat-table/flat-table.interaction.stories.tsx (hunk 10 above)
Comment (sianford) [reply src/components/flat-table/flat-table.interaction.stories.tsx]: I've removed it as the table container in the Pagination example is also focused so it's not required.

File: src/components/flat-table/flat-table.interaction.stories.tsx (hunk 12 above)
Comment (sianford) [reply src/components/flat-table/flat-table.interaction.stories.tsx:211]: I'll create a follow up ticket to investigate the focus cases as I tried to focus many different elements with `pseudo` and couldn't get them to work for all required elements.

File: src/components/flat-table/flat-table.interaction.stories.tsx (hunk 13 above)
Comment (sianford) [reply src/components/flat-table/flat-table.interaction.stories.tsx:365]: Also, I removed lots of snapshots from the various examples in the main stories so I combined focusing these elements with snapshotting.

File: src/components/flat-table/flat-table.pw.tsx
Comment (sianford) [reply src/components/flat-table/flat-table.pw.tsx:1]: I left in mostly only those tests that weren't covered in unit or in Chromatic. They're all more complex scenarios, for instance checking the scrolling up and down the table content when `hasStickyHead` is set, boundary value analysis on different heights with `minHeight` set, checking the full character diacritics, keyboard accessibility using space/Enter/arrows etc that weren't covered by unit tests. After another pass, I've incorporated some keyboard accessibility tests into the interactions tests and removed them from here.

File: src/components/flat-table/flat-table.pw.tsx (hunk 5 above)
Comment (sianford) [reply src/components/flat-table/flat-table.pw.tsx]: Added a story for rounded corners for Chromatic to capture and removed these tests.

File: src/components/flat-table/flat-table.pw.tsx (hunk 6 above)
Comment (sianford) [reply src/components/flat-table/flat-table.pw.tsx]: Removed

File: src/components/flat-table/flat-table.pw.tsx (hunk 7 above)
Comment (sianford) [reply src/components/flat-table/flat-table.pw.tsx]: The thinking was it might highlight a problem with overlapping text. Removed though.

**Review by paulrobinson88 — APPROVED** [review]
LGTM, as you've mentioned it might be worth another pass at this at some point but that's a good chunk of redundant tests removed and snapshots refined, nice work.

File: src/components/flat-table/flat-table.pw.tsx (hunk 8 above)
Comment (edleeks87) [reply src/components/flat-table/flat-table.pw.tsx:1395]: Axe won't flag anything to do with spacing in the cells, it might catch if the borders overlay (incomplete contrast etc) and but they're all the same colour the table semantics shouldn't allow for that

File: src/components/flat-table/flat-table.interaction.stories.tsx (hunk 10 above)
Comment (edleeks87) [reply src/components/flat-table/flat-table.interaction.stories.tsx]: the root is, the rest of the table isn't

**Review by edleeks87 — APPROVED** [review]
Approving but agree this needs a further pass
